specs/BrowserProcessExited.md

```diff
@@ -0,0 +1,322 @@
+        [browserExitedEventToken, this](
+            ICoreWebView2Environment* sender,
+            ICoreWebView2BrowserProcessExitedEventArgs* args) {
+        COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND kind;
+        CHECK_FAILURE(args->get_BrowserProcessExitKind(&kind));
+
+        // Watch for graceful browser process exit. Let ProcessFailed event
+        // handler take care of failed browser process termination.
+        if (kind == COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND_NORMAL)
+        {
+          CHECK_FAILURE(
+              m_webViewEnvironment->remove_BrowserProcessExited(browserExitedEventToken));
+          // Release the environment only after the handler is invoked.
+          // Otherwise, there will be no environment to raise the event when
+          // the process exits.
+          m_webViewEnvironment = nullptr;
+          CleanupUserDataFolder();
+        }
+
+        return S_OK;
+    }).Get(),
+    &browserExitedEventToken));
+```
+
+## .NET C#
+```c#
+// URI or other state to save/restore when the WebView is recreated.
+private Uri _uriToRestore;
+
+async void RegisterForNewVersion()
+{
+    // We need to make sure the CoreWebView2 property is not null, so we can get
+    // the environment from it. Alternatively, if the WebView was created from
+    // an environment provided to the control, we can use that environment
+    // object directly.
+    await webView.EnsureCoreWebView2Async();
+    _coreWebView2Environment = webView.CoreWebView2.Environment;
+    _coreWebView2Environment.NewBrowserVersionAvailable += Environment_NewBrowserVersionAvailable;
+}
+
+// A new version of the WebView2 Runtime is available, our handler gets called.
+// We close our WebView and set a handler to reinitialize it once the browser
+// process is gone, so we get the new version of the WebView2 Runtime.
+void Environment_NewBrowserVersionAvailable(object sender, object e)
+{
+    StringBuilder messageBuilder = new StringBuilder(256);
+    messageBuilder.Append("We detected there is a new version of the WebView2 Runtime installed. ");
+    messageBuilder.Append("Do you want to switch to it now? This will re-create the WebView.");
+    var selection = MessageBox.Show(this, messageBuilder.ToString(), "New WebView2 Runtime detected", MessageBoxButton.YesNo);
+    if (selection == MessageBoxResult.Yes)
+    {
+        // Save URI or other state you want to restore when the WebView is recreated.
+        _uriToRestore = webView.Source;
+        _coreWebView2Environment.BrowserProcessExited += Environment_BrowserProcessExited;
+        // We dispose of the control so the internal WebView objects are released
+        // and the associated browser process exits. If there are any other WebViews
+        // from the same environment configuration, they need to be closed too.
+        webView.Dispose();
+        webView = null;
+    }
+}
+
+void Environment_BrowserProcessExited(object sender, CoreWebView2BrowserProcessExitedEventArgs e)
```

Comment:
`sender` should be declared as `CoreWebView2Environment`. This also removes the need to cast it below.

Reply:
Thanks. Updated.